core/shared/src/main/scala/org/typelevel/log4cats/AskStructuredLogger.scala

```diff
@@ -0,0 +1,71 @@
+package org.typelevel.log4cats
+
+import cats.Monad
+import cats.mtl.Ask
+
+private[log4cats] class AskStructuredLogger[F[_]](logger: StructuredLogger[F])(implicit
+    F: Monad[F],
+    ask: Ask[F, Map[String, String]]
+) extends StructuredLogger[F] {
+
+  override def error(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.error(askCtx)(message))
+
+  override def warn(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.warn(askCtx)(message))
+
+  override def info(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.info(askCtx)(message))
+
+  override def debug(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.debug(askCtx)(message))
+
+  override def trace(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.trace(askCtx)(message))
+
+  override def error(t: Throwable)(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.error(askCtx, t)(message))
+
+  override def warn(t: Throwable)(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.warn(askCtx, t)(message))
+
+  override def info(t: Throwable)(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.info(askCtx, t)(message))
+
+  override def debug(t: Throwable)(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.debug(askCtx, t)(message))
+
+  override def trace(t: Throwable)(message: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.trace(askCtx, t)(message))
+
+  override def trace(ctx: Map[String, String])(msg: => String): F[Unit] =
+    F.flatMap(ask.ask)(askCtx => logger.trace(askCtx ++ ctx)(msg))
```

Comment:
Performance concerns, not yet proven by benchmarking:

1. We pay to ask and flatMap when the level is disabled.
2. In the methods with a `ctxt`, we additionally pay to concatenate askCtx and ctx when the level is disabled.

An underlying SelfAwareLogger could help, but then we'd be an AskSelfAwareStructuredLogger.  The traits aren't composing particularly well.

We could further optimize by getting into `Slf4jLogger`, but that may further contribute to the combinatorial explosion.

Reply:
as annoying as it would be, I don't think a single "duplicate" for `SelfAwareStructuredLogger` is *too* awful